### ISSUE TITLE
AutoGluon WSL2 CUDA Crash Bug Report

### ISSUE THREAD

**Issue by Fsaupe** [awslabs/autogluon#5187]
**Describe the bug**
AutoGluon training crashes during dynamic stacking phase in WSL2 environment when FastAI models attempt CUDA device detection via torch.cuda.is_available(). The crash occurs consistently during the leaderboard generation phase, specifically when FastAI models try to determine the default device.

   - Training actually succeeds (87.4% accuracy achieved)
   - Crash happens during cleanup/evaluation phase
   - Non-FastAI models work perfectly in this environment
   - Issue appears specific to FastAI + Ray + WSL2 CUDA interaction

**Expected behavior**
Training should conclude successfully

**Reproduction Steps**

   AutoGluon Configuration

   predictor = TabularPredictor(
       label='Kategorie',
       path='autogluon_model',
       problem_type='multiclass',
       eval_metric='accuracy'
   )

   predictor.fit(
       train_data,
       time_limit=360,  # 6 minutes
       presets='best_quality',
       holdout_frac=0.1,
       verbosity=4,
       num_gpus=1,
       ag_args_fit={'num_gpus': 1, 'max_memory_usage_ratio': 2}
   )

   Dataset

   - Rows: 13,529 (after cleaning)
   - Features: 21 (mix of numerical, categorical, and text features)
   - Target Classes: 87 unique categories
   - Problem Type: Multiclass classification

   Crash Pattern

   1. Training proceeds normally through L1 models (KNN, FastAI, LightGBM, etc.)
   2. Many L2 models get skipped due to time constraints
   3. Training completes successfully with 87.4% validation accuracy
   4. Crash occurs during dynamic stacking completion when generating leaderboard
   5. FastAI models crash specifically during torch.cuda.is_available() call

   Root Cause Analysis

   Technical Details

   - Issue: WSL2 CUDA driver stub has memory management conflicts with Ray multiprocessing
   - Trigger: FastAI's device detection via torch.cuda.device_count() in Ray worker processes
   - Environment: Specific to WSL2 virtualized CUDA environment
   - Timing: Occurs during model evaluation/leaderboard generation phase

   Memory Corruption Pattern

   free(): double free detected in tcache 2
   cuInit -> cudaGetDeviceCount -> c10::cuda::device_count()

   Workarounds Tested

   Failed Attempts

   1. CUDA Toolkit Update: Updated from 12.6 → 12.9 (no improvement)

   Successful Workaround

   predictor.fit(
       train_data,
       excluded_model_types=['FASTAI'],  # Exclude FastAI models
       num_gpus=1,
       ag_args_fit={'num_gpus': 1, 'max_memory_usage_ratio': 2} //strange that i have to use num_gpus 2 times, without num_gpus in ag_args_fit GPU is not used
   )

   System Information

   - OS: Ubuntu 24.04.1 LTS (Noble Numbat) running in WSL2
   - WSL2 Kernel: 5.15.167.4-microsoft-standard-WSL2
   - Platform: Windows Subsystem for Linux 2
   - Python: 3.12.3

   GPU/CUDA Setup

   - GPU: NVIDIA GeForce RTX 4070 Ti
   - Windows NVIDIA Driver: 576.52
   - WSL2 NVIDIA Stub Driver: 575.57.04
   - CUDA Toolkit: 12.9.86 (latest, installed May 2025)
   - CUDA Runtime: 12.9
  
   Package Versions

   - AutoGluon Tabular: 1.3.1
   - PyTorch: 2.6.0+cu124
   - Ray: 2.44.1
   - FastAI: 2.8.2
   - Pandas: 2.2.3
   - NumPy: 2.1.3
   - Scikit-learn: 1.6.1

**Comment by Fsaupe** [comment]
after i had a successful run yesterday, it now crashed with XGBoost, too: 

## Describe the bug

AutoGluon training crashes with CUDA memory corruption and segmentation fault during final model evaluation phase. The training completes successfully (achieving 87.56% accuracy with WeightedEnsemble_L3) but crashes when loading XGBoost models during the final evaluation step.

## To Reproduce

### Steps to reproduce the behavior:

1. Set up AutoGluon with GPU support (CUDA enabled)
2. Train a multiclass classification model on tabular data with the following configuration:
   ```python
   predictor = TabularPredictor(
       label='Kategorie',
       path='autogluon_model',
       problem_type='multiclass',
       eval_metric='accuracy'
   )
   
   predictor.fit(
       train_data,
       time_limit=3600,
       presets='best_quality',
       holdout_frac=0.1,
       verbosity=4,
       num_gpus=1,
       excluded_model_types=['FASTAI'],
       ag_args_fit={'num_gpus': 1, 'max_memory_usage_ratio': 2}
   )
   ```
3. Train on a dataset with ~8000 rows, 68 unique categories, with feature engineering creating ~30 features
4. Allow training to complete normally (training succeeds)
5. Crash occurs during final evaluation when `predictor.evaluate(test_data)` is called

### Code snippet

**Training script (train_autogluon.py):**
```python
#!/usr/bin/env python3
"""
Train AutoGluon model on transaction data for category classification
"""

import pandas as pd
import numpy as np
from autogluon.tabular import TabularDataset, TabularPredictor
from sklearn.model_selection import train_test_split
import os
import argparse

def train_model(df, model_path='autogluon_model', test_size=0.1, time_limit=3600):
    """Train the AutoGluon model"""
    print("Preparing training data...")
    
    # Split the data
    train_data, test_data = train_test_split(df, test_size=test_size, random_state=42, stratify=df['Kategorie'])
    
    print(f"Training set: {len(train_data)} rows")
    print(f"Test set: {len(test_data)} rows")
    
    # Convert to TabularDataset
    train_data = TabularDataset(train_data)
    test_data = TabularDataset(test_data)
    
    # Create predictor
    print("Training AutoGluon model...")
    predictor = TabularPredictor(
        label='Kategorie',
        path=model_path,
        problem_type='multiclass',
        eval_metric='accuracy'
    )
    
    # Train the model
    predictor.fit(
        train_data,
        time_limit=time_limit,
        presets='best_quality',
        holdout_frac=0.1,
        verbosity=4,
        num_gpus=1,
        excluded_model_types=['FASTAI'],
        ag_args_fit={'num_gpus': 1,'max_memory_usage_ratio': 2}
    )
    
    # Evaluate on test set - THIS IS WHERE THE CRASH OCCURS
    print("Evaluating model...")
    predictions = predictor.predict(test_data)
    performance = predictor.evaluate(test_data)  # CRASH HERE
    
    return predictor, performance
```

### Expected behavior

Training should complete successfully including the final evaluation step without memory corruption or crashes.

### Actual behavior

Training proceeds normally for 902.4 seconds and achieves good performance (87.56% accuracy), but crashes during final evaluation with memory corruption in XGBoost model loading.

## Environment

### Environment info

- **OS**: Linux 5.15.167.4-microsoft-standard-WSL2 (WSL2 on Windows)
- **Python version**: 3.x
- **AutoGluon version**: Latest (as of 2025-07-07)
- **GPU**: CUDA-enabled GPU
- **Ray version**: Distributed training enabled

### Package versions
```
autogluon (latest)
pandas
numpy
sklearn
xgboost
ray (for distributed training)
CUDA toolkit
```

## Error messages and logs

### Main Error Output:
```
Evaluating model...
(_dystack pid=83153) Skipping NeuralNetTorch_r185_BAG_L2 as files are corrupted
(_dystack pid=83153) Loading pretrained NeuralNetTorch_r45 model...
(_dystack pid=83153) Loaded model in 1.46s.
(_dystack pid=83153) Skipping XGBoost_r8_BAG_L2 as files are corrupted
(_dystack pid=83153) Loading pretrained XGBoost_r176 model...
free(): double free detected in tcache 2
(_dystack pid=83153) *** SIGABRT received at time=1720373669 on cpu 6 ***
(_dystack pid=83153) PC: @     0x7f7f4fd3000c (unknown)  gsignal
(_dystack pid=83153)     @     0x7f7f4fcf8472 (unknown)  __libc_start_main
(_dystack pid=83153)     @     0x7f7f4fd301fb (unknown)  (unknown)
```

### Training Progress Before Crash:
```
Fitting model: WeightedEnsemble_L3...
Ensemble Weights: {'NeuralNetTorch_r185_BAG_L2': 0.111, 'XGBoost_r176_BAG_L2': 0.889}
WeightedEnsemble_L3  = 0.8756    | 0.82s | (Ensemble)
AutoGluon training complete, total runtime = 902.40s ... Best model: "WeightedEnsemble_L3"
TabularPredictor saved. To load, use: TabularPredictor.load("autogluon_model")
